casepro/backend/rapidpro.py

```diff
@@ -253,29 +253,39 @@ def create_label(self, org, name):
 
         return remote.uuid
 
-    def push_outgoing(self, org, outgoing):
+    def push_outgoing(self, org, outgoing, as_broadcast=False):
         client = self._get_client(org, 1)
-        contact_uuids = [c.uuid for c in outgoing.contacts.all()]
-        broadcast = client.create_broadcast(text=outgoing.text, contacts=contact_uuids, urns=list(outgoing.urns))
 
-        outgoing.backend_id = broadcast.id
-        outgoing.save(update_fields=('backend_id',))
+        if as_broadcast:
+            contact_uuids = []
+            urns = []
+            for msg in outgoing:
+                if msg.contact:
+                    contact_uuids.append(msg.contact.uuid)
+                if msg.urns:
+                    urns.extend(msg.urns)
+            text = outgoing[0].text
+            broadcast = client.create_broadcast(text=text, contacts=contact_uuids, urns=urns)
+
+            for msg in outgoing:
+                msg.backend_broadcast_id = broadcast.id
+
+            Outgoing.objects.filter(pk__in=[o.id for o in outgoing]).update(backend_broadcast_id=broadcast.id)
+        else:
+            for msg in outgoing:
+                contact_uuids = [msg.contact.uuid] if msg.contact else []
+                broadcast = client.create_broadcast(text=msg.text, contacts=contact_uuids, urns=msg.urns)
+
+                msg.backend_broadcast_id = broadcast.id
+                msg.save(update_fields=('backend_broadcast_id',))
 
     def add_to_group(self, org, contact, group):
         client = self._get_client(org, 1)
-        # TODO until we know which groups are dynamic, we should assume this call might fail
-        try:
-            client.add_contacts([contact.uuid], group_uuid=group.uuid)
-        except Exception:
-            pass
+        client.add_contacts([contact.uuid], group_uuid=group.uuid)
```

Comment:
We now do know which groups are dynamic and these methods won't be called for dynamic groups